File: hpdic/script/benchmark_baseline.py
import time
import sys
import os
import numpy as np
import faiss

# ================= 配置区域 =================
DATA_DIR = os.path.expanduser("~/ElasticIVF/hpdic/data/sift")
BASE_FILE = os.path.join(DATA_DIR, "sift_base.fvecs")
LEARN_FILE = os.path.join(DATA_DIR, "sift_learn.fvecs")

DIM = 128
NLIST = 1024          # 聚类中心数量
WINDOW_SIZE = 100000  # 窗口大小 10w
BATCH_SIZE = 10000    # 每次滑动 1w
DEVICE_ID = 0         # RTX 6000 ID

# ================= 工具函数 =================
def ivecs_read(fname):
    a = np.fromfile(fname, dtype='int32')
    d = a[0]
    return a.reshape(-1, d + 1)[:, 1:].copy()

def fvecs_read(fname):
    return ivecs_read(fname).view('float32')

def get_memory_usage():
    # 简单的显存占用估算 (仅作参考)
    try:
        import pynvml
        pynvml.nvmlInit()
        handle = pynvml.nvmlDeviceGetHandleByIndex(DEVICE_ID)
        info = pynvml.nvmlDeviceGetMemoryInfo(handle)
        return info.used / 1024**2
    except:
        return 0

# ================= 主流程 =================
def main():
    print(f"Loading data from {DATA_DIR}...")
    if not os.path.exists(BASE_FILE):
        print(f"Error: {BASE_FILE} not found.")
        return

    xb = fvecs_read(BASE_FILE)
    xt = fvecs_read(LEARN_FILE)
    print(f"Data loaded. Base shape: {xb.shape}")

    # 1. 初始化 GPU
    print("Initializing Faiss GPU resources...")
    res = faiss.StandardGpuResources()
    
    # 2. 构建并训练索引 (使用 IVFFlat)
    print("Training IndexIVFFlat on CPU...")
    quantizer = faiss.IndexFlatL2(DIM)
    cpu_index = faiss.IndexIVFFlat(quantizer, DIM, NLIST, faiss.METRIC_L2)
    cpu_index.train(xt)

    # 3. 搬运到 GPU
    print("Moving index to GPU...")
    # co = faiss.GpuClonerOptions()
    # co.useFloat16 = True # 如果需要省显存可以开启
    gpu_index = faiss.index_cpu_to_gpu(res, DEVICE_ID, cpu_index)

    # 4. 预填充窗口
    print(f"Pre filling window with {WINDOW_SIZE} vectors...")
    gpu_index.add(xb[:WINDOW_SIZE])
    
    # 记录当前在索引里的 IDs (Faiss 默认 ID 是连续的 0..N)
    # 我们用一个简单的 FIFO 队列逻辑来模拟 ID
    current_min_id = 0
    current_max_id = WINDOW_SIZE
    
    print("\n>>> Starting Streaming Benchmark <<<")
    print(f"Window: {WINDOW_SIZE}, Batch: {BATCH_SIZE}")
    print(f"{'Step':<5} | {'Add(ms)':<10} | {'Remove(ms)':<12} | {'Method':<15} | {'Total(ms)':<10}")
    print("="*65)

    # 运行 10 轮即可看出趋势
    for step in range(10):
        # 准备数据
        start_idx = WINDOW_SIZE + step * BATCH_SIZE
        if start_idx + BATCH_SIZE > xb.shape[0]:
            break
            
        new_data = xb[start_idx : start_idx + BATCH_SIZE]
        
        # 这一批要删除的 IDs (模拟最老的 ID)
        ids_to_remove = np.arange(current_min_id, current_min_id + BATCH_SIZE, dtype=np.int64)

        # A. 测试 Add
        t0 = time.time()
        gpu_index.add(new_data)
        t1 = time.time()
        add_time = (t1 - t0) * 1000

        # B. 测试 Remove
        t2 = time.time()
        method_name = "Direct"
        try:
            # 尝试直接在 GPU 上删除
            gpu_index.remove_ids(ids_to_remove)
        except Exception as e:
            # 如果报错 (NotImplemented)，则模拟 "CPU Roundtrip"
            # 这是目前 Faiss 用户被迫使用的慢速方案
            method_name = "CPU_Roundtrip"
            
            # 1. 把索引拷回 CPU
            tmp_cpu_index = faiss.index_gpu_to_cpu(gpu_index)
            # 2. 在 CPU 上删除
            tmp_cpu_index.remove_ids(ids_to_remove)
            # 3. 重新拷回 GPU (极慢)
            gpu_index = faiss.index_cpu_to_gpu(res, DEVICE_ID, tmp_cpu_index)
            
        t3 = time.time()
        remove_time = (t3 - t2) * 1000
        
        total_time = add_time + remove_time
        
        print(f"{step:<5} | {add_time:<10.2f} | {remove_time:<12.2f} | {method_name:<15} | {total_time:<10.2f}")

        # 更新 ID 计数
        current_min_id += BATCH_SIZE
        current_max_id += BATCH_SIZE

if __name__ == "__main__":
    main()

File: hpdic/experiment/benchmark_baseline.cpp
/**
 * benchmark_baseline.cpp
 *
 * Author: Dongfang Zhao
 * Email:  dzhao@uw.edu
 *
 * Benchmark for Add and Remove operations on Faiss GPU IndexIVFFlat.
 * * This script simulates a sliding window workload:
 * 1. Adds a new batch of vectors.
 * 2. Removes the oldest batch of vectors.
 * * Since standard Faiss GPU indices do not support native removal, this benchmark
 * captures the cost of the "CPU Roundtrip" workaround (GPU -> CPU -> Remove -> GPU).
 */

#include <cstdio>
#include <vector>
#include <chrono>
#include <iostream>
#include <numeric>

#include <faiss/IndexFlat.h>
#include <faiss/IndexIVFFlat.h>
#include <faiss/gpu/GpuIndexIVFFlat.h>
#include <faiss/gpu/StandardGpuResources.h>
#include <faiss/gpu/GpuCloner.h>

using namespace std;
using namespace std::chrono;

/**
 * Helper: Read fvecs file format.
 * Format: [dim][v1...][dim][v2...]
 */
float* fvecs_read(const char* fname, size_t& d, size_t& n) {
    FILE* f = fopen(fname, "rb");
    if (!f) {
        perror("Failed to open file");
        return nullptr;
    }

    // 1. Read the dimension from the header of the first vector
    int d_int;
    if (fread(&d_int, sizeof(int), 1, f) != 1) {
        fprintf(stderr, "Error reading dimension from %s\n", fname);
        fclose(f);
        return nullptr;
    }
    d = (size_t)d_int;

    // 2. Calculate the number of vectors (n) based on file size
    fseek(f, 0, SEEK_END);
    size_t file_size = ftell(f);
    rewind(f);

    // Each vector contains: 1 int (4 bytes) + d floats (4 * d bytes)
    size_t entry_size = sizeof(int) + d * sizeof(float);

    if (entry_size == 0 || file_size % entry_size != 0) {
        fprintf(stderr, "Error: File size is not a multiple of vector size.\n");
        fclose(f);
        return nullptr;
    }

    n = file_size / entry_size;

    // 3. Allocate memory and read data
    float* data = new float[n * d];

    for (size_t i = 0; i < n; i++) {
        // Skip the dimension header (int) for the current vector
        if (fseek(f, sizeof(int), SEEK_CUR) != 0) {
            perror("fseek failed");
            delete[] data;
            fclose(f);
            return nullptr;
        }

        // Read the float data
        if (fread(data + i * d, sizeof(float), d, f) != d) {
            fprintf(stderr, "Error reading vector %zu\n", i);
            delete[] data;
            fclose(f);
            return nullptr;
        }
    }

    fclose(f);
    return data;
}

int main() {
    // ==========================================
    // Configuration Parameters
    // ==========================================
    const char* base_file = "/home/cc/ElasticIVF/hpdic/data/sift/sift_base.fvecs";
    const char* learn_file = "/home/cc/ElasticIVF/hpdic/data/sift/sift_learn.fvecs";
    int nlist = 1024;
    int window_size = 100000;
    int batch_size = 10000;
    int device_id = 0;

    // 1. Load Data
    size_t d, nb, nt;
    float* xb = fvecs_read(base_file, d, nb);
    float* xt = fvecs_read(learn_file, d, nt);
    printf("Data loaded. Dim: %zu, Base: %zu\n", d, nb);

    // 2. Initialize GPU Resources
    faiss::gpu::StandardGpuResources res;

    // 3. Build and Train Index (on CPU first)
    faiss::IndexFlatL2 quantizer(d);
    faiss::IndexIVFFlat cpu_index(&quantizer, d, nlist, faiss::METRIC_L2);
    
    printf("Training index...\n");
    cpu_index.train(nt, xt);

    // 4. Move Index to GPU
    printf("Moving index to GPU...\n");
    faiss::gpu::GpuIndexIVFFlatConfig config;
    config.device = device_id;
    faiss::gpu::GpuIndexIVFFlat gpu_index(&res, &cpu_index, config);

    // 5. Pre-fill Window
    printf("Pre-filling window with %d vectors...\n", window_size);
    gpu_index.add(window_size, xb);

    long current_min_id = 0;

    printf("\n%-5s | %-10s | %-12s | %-15s | %-10s\n", "Step", "Add(ms)", "Remove(ms)", "Method", "Total(ms)");
    printf("-----------------------------------------------------------------\n");

    for (int step = 0; step < 10; step++) {
        size_t start_idx = window_size + step * batch_size;
        float* new_data = xb + start_idx * d;

        // Prepare IDs to remove (simulating FIFO queue)
        vector<faiss::idx_t> ids_to_remove(batch_size);
        iota(ids_to_remove.begin(), ids_to_remove.end(), current_min_id);

        // A. Benchmark Add
        auto t0 = high_resolution_clock::now();
        gpu_index.add(batch_size, new_data);
        auto t1 = high_resolution_clock::now();
        double add_time = duration<double, milli>(t1 - t0).count();

        // B. Benchmark Remove
        auto t2 = high_resolution_clock::now();
        string method_name = "Direct";
        
        try {
            // Attempt direct GPU removal (Expected to fail on Standard Faiss)
            faiss::IDSelectorBatch selector(batch_size, ids_to_remove.data());
            gpu_index.remove_ids(selector);
        } catch (const exception& e) {
            // Fallback: Simulate "CPU Roundtrip" overhead
            method_name = "CPU_Roundtrip";
            
            // 1. Download: GPU to CPU
            faiss::IndexIVFFlat* tmp_cpu_index = dynamic_cast<faiss::IndexIVFFlat*>(
                faiss::gpu::index_gpu_to_cpu(&gpu_index)
            );
            
            // 2. Modify: CPU Remove
            faiss::IDSelectorBatch selector(batch_size, ids_to_remove.data());
            tmp_cpu_index->remove_ids(selector);
            
            // 3. Upload: CPU to GPU (Reconstruct GPU index to simulate full overhead)
            gpu_index.~GpuIndexIVFFlat(); // Explicitly destroy old GPU index
            new (&gpu_index) faiss::gpu::GpuIndexIVFFlat(&res, tmp_cpu_index, config); // Placement new
            
            delete tmp_cpu_index;
        }
        
        auto t3 = high_resolution_clock::now();
        double remove_time = duration<double, milli>(t3 - t2).count();

        printf("%-5d | %-10.2f | %-12.2f | %-15s | %-10.2f\n", 
               step, add_time, remove_time, method_name.c_str(), add_time + remove_time);

        current_min_id += batch_size;
    }

    delete[] xb;
    delete[] xt;
    return 0;
}

/**
 * Compilation Command:
 * * g++ -O3 -std=c++17 -fopenmp benchmark_baseline.cpp -o benchmark_baseline.bin \
 * -I/home/cc/ElasticIVF \
 * -I/usr/local/cuda/include \
 * -L/home/cc/ElasticIVF/build/faiss \
 * -L/usr/local/cuda/lib64 \
 * -lfaiss \
 * -lopenblas \
 * -lcudart \
 * -lcublas
 */